### PR TITLE
[onert] Enable max2d PoolLayer for training

### PR DESCRIPTION
**Pull request by zetwhite**
This PR is to enable max2d PoolLayer for training.

This PR fills unimplemented part of PoolLayer: 

  * implements shape-inference visit() for pool2d
  * implelments kernel-generator visit() for pool2d
  * Implements train PoolLayer op's forward, backward functions and ckernel
  
ONE-DCO-1.0-Signed-off-by: SeungHui Youn <sseung.youn@samsung.com> 

---------------------------------------------------

related issue : https://github.com/Samsung/ONE/issues/11427 
from draft : https://github.com/Samsung/ONE/pull/11480

### CODE REVIEW

File: runtime/onert/backend/cpu/ops/PoolLayer.h [hunk 1]
```diff
@@ -53,10 +53,11 @@ class PoolLayer : public ::onert::exec::IFunction
 
   void run() override;
 
-private:
+protected:
   const IPortableTensor *_input;
   IPortableTensor *_output;
```
Comment (zetwhite) [inline runtime/onert/backend/cpu/ops/PoolLayer.h:58]: These need to be accessible by `train/op/PoolLayer`.

File: runtime/onert/backend/train/KernelGenerator.h [hunk 2]
```diff
@@ -50,9 +50,11 @@ class KernelGenerator : public backend::train::KernelGeneratorBase
   void visit(const ir::train::operation::ElementwiseActivation &) override;
   void visit(const ir::train::operation::FullyConnected &) override;
   void visit(const ir::train::operation::Loss &) override;
+  void visit(const ir::train::operation::Pool2D &) override;
   void visit(const ir::train::operation::Reshape &node) override;
 
 private:
+  const ir::Operands &_ctx;
```
Comment (zetwhite) [inline runtime/onert/backend/train/KernelGenerator.h:57]: This is necessary to get `pool layer`'s input and output tensor shape.
In/Out shape is used to calculate the pool layer's padding.

File: runtime/onert/backend/train/ops/PoolLayer.cc [hunk 3]
```diff
@@ -25,7 +31,104 @@ namespace train
...
+  switch (pool_type)
+  {
+    case train::ops::PoolType::kMax:
+      return cpu::ops::PoolType::kMax;
+    default:
+      throw std::runtime_error("PoolLayer: Unsupported pool type yet");
+  }
+}
+
+class MaxPool2D final : public TrainingKernelRegistry
```
Comment (zetwhite) [inline runtime/onert/backend/train/ops/PoolLayer.cc:48]: This class is a wrapper of  "max pool" related things. 

It provides
	* forward and backward interfaces for maxpooling (call proper kernel) 
	* holds additional memory used to record the max-arguments' index.

File: runtime/onert/backend/train/ops/PoolLayer.cc [hunk 4]
```diff
@@ -35,14 +147,32 @@ void PoolLayer::configure(const IPortableTensor *input, const uint32_t paddingLe
...
+  cpu::ops::PoolLayer::configure(input, paddingLeft, paddingRight, paddingTop, paddingBottom,
+                                 strideWidth, strideHeight, kernelWidth, kernelHeight, activation,
+                                 output, convert_to_infer(op_type));
+
+  if (output->data_type() != OperandType::FLOAT32)
+  {
+    throw std::runtime_error("PoolLayer : Unsupported data type for training");
+  }
+
+  // ready training kernel
```
Comment (zetwhite) [inline runtime/onert/backend/train/ops/PoolLayer.cc:169]: Both Inference and training kernel is ready in compile time.
In Runtime, depends on `bool training` value, It decides which kernel to use.

File: compute/cker/include/cker/train/operation/MaxPool.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+// Most of the logic except 'arg_max_index' related is copy-paste from
+// https://github.com/Samsung/ONE/blob/a380292/compute/cker/include/cker/operation/MaxPool.h#L42-L88
+// 'arg_max_index' is to record max-arguments' index to apply gradient later.
+inline void MaxPool2D(const PoolParams &params, const Shape &input_shape, const float *input_data,
+                      const Shape &output_shape, float *output_data, int *arg_max_index)
+{
+  assert(input_shape.DimensionsCount() == 4);
+  assert(output_shape.DimensionsCount() == 4);
+  assert(input_shape.Dims(3) == output_shape.Dims(3)); // MaxPool2D doesn't change depth
```
Comment (ragmani) [inline compute/cker/include/cker/train/operation/MaxPool.h:41]: (Optional) How about adding to check batches?
```suggestion
  assert(input_shape.Dims(0) == output_shape.Dims(0)); // MaxPool2D doesn't change batch
  assert(input_shape.Dims(3) == output_shape.Dims(3)); // MaxPool2D doesn't change depth
```

File: runtime/onert/backend/train/KernelGenerator.h (hunk 2 above)
Comment (ragmani) [reply runtime/onert/backend/train/KernelGenerator.h:57]: (Optional) How about using the base member `_tgraph` instead of adding this member (because performance is not important where this member is used) ?

File: runtime/onert/backend/train/ops/PoolLayer.cc [hunk 6]
```diff
@@ -25,7 +31,113 @@ namespace train
 namespace ops
 {
 
-PoolLayer::PoolLayer() : cpu::ops::PoolLayer()
+namespace
+{
+
+cpu::ops::PoolType convert_to_infer(const train::ops::PoolType &pool_type)
```
Comment (ragmani) [inline runtime/onert/backend/train/ops/PoolLayer.cc:37]: (Optional)
```suggestion
cpu::ops::PoolType convertToInfer(const train::ops::PoolType &pool_type)
```
Although there are many places in onert that did not follow naming conventions, it would be better to use camel case that starts with lowercase.

File: compute/cker/include/cker/train/operation/MaxPool.h (hunk 5 above)
Comment (jyoungyun) [reply compute/cker/include/cker/train/operation/MaxPool.h:41]: The first dimension is already being checked in the `MatchingDim` function below. So this overlaps. :)

File: runtime/onert/backend/train/ops/PoolLayer.cc [hunk 7]
```diff
@@ -25,7 +31,113 @@ namespace train
...
+    // maxpool forward
+    nnfw::cker::train::MaxPool2D(_op_params, getShape(in), getBuffer<float>(in), out_shape,
+                                 out_data, getBuffer<int>(arg_max_index));
+
+    // activation forward
+    for (int i = 0; i < out_shape.FlatSize(); ++i)
+    {
+      out_data[i] = nnfw::cker::ActivationFunctionWithMinMax(
+        out_data[i], _op_params.float_activation_min, _op_params.float_activation_max);
+    }
```
Comment (jyoungyun) [inline runtime/onert/backend/train/ops/PoolLayer.cc:106]: (Q) Why do you implement this activation function here? The original MaxPool.h kernel code implements this function inside the cker. Could you explain your intention?

File: compute/cker/include/cker/train/operation/MaxPool.h [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  const int input_width = input_shape.Dims(2);
+  const int output_height = output_shape.Dims(1);
+  const int output_width = output_shape.Dims(2);
+  const int stride_height = params.stride_height;
+  const int stride_width = params.stride_width;
+
+  const auto in_mat = MapAsMatrixWithLastDimAsRows(input_data, input_shape);
+  auto out_mat = MapAsMatrixWithLastDimAsRows(output_data, output_shape);
+  auto arg_max_index_mat = MapAsMatrixWithLastDimAsRows(arg_max_index, output_shape);
+
```
Comment (jyoungyun) [inline compute/cker/include/cker/train/operation/MaxPool.h:54]: Could you explain the grad formula for MaxPool function [here](https://github.com/Samsung/ONE/issues/11248)? It will help me understand your code. :)

File: runtime/onert/backend/train/ops/PoolLayer.cc (hunk 7 above)
Comment (zetwhite) [reply runtime/onert/backend/train/ops/PoolLayer.cc:106]: I hope that `forward` and `backward` functions are matched somehow. 
Since the `backward` explicitly calls gradient-calculation function for `maxpool` and `relu`, I hope that `forward` has a similar look.  

I also thought that if `avgpool` is introduced later, this activation-related part could be exported to a common-thing (to share between `avgpool` and `maxpool`).
To export this common part, I thought it might better to have logic explicit here rather than inside the cker.

Comment (zetwhite) [reply runtime/onert/backend/train/ops/PoolLayer.cc:106]: But I don't have a strong opinion about this part. 
So, If you think it might be better to move inside the `cker`, please let me know. :smile:

File: compute/cker/include/cker/train/operation/MaxPool.h [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            // update out_mat
+            out_mat.col(out_offset) = out_vector.cwiseMax(in_vector);
+          }
+        }
+      }
+    }
+  }
+}
+
+inline void MaxPool2DGrad(const Shape &deriv_output_shape, const float *deriv_output_data,
```
Comment (zetwhite) [inline compute/cker/include/cker/train/operation/MaxPool.h:102]: Additionally also here, `MaxPool2DGrad` only do maxpool thing, not activation related. 
So, To make corresponded `MaxPool2D`, I thought it looks more natural that `MaxPool2D` only does max pooling, not activation.

File: runtime/onert/backend/train/ops/PoolLayer.cc (hunk 7 above)
Comment (zetwhite) [reply runtime/onert/backend/train/ops/PoolLayer.cc:106]: [Additionally, Also here ](https://github.com/Samsung/ONE/pull/11686/files/4c2d6bbd23ce01ef3c3212bc213045fb5d7a5ab8#r1351950822)

File: compute/cker/include/cker/train/operation/MaxPool.h (hunk 5 above)
Comment (ragmani) [reply compute/cker/include/cker/train/operation/MaxPool.h:41]: I think it's not bad because it doesn't affect performance.

File: compute/cker/include/cker/train/operation/MaxPool.h (hunk 9 above)
Comment (glistening) [reply compute/cker/include/cker/train/operation/MaxPool.h:102]: @zetwhite We handles several types (e.g. max, avg, ...) of pooling in `PoolLayer`. It handles activation (including no activation) using `*_activation_min/max` in parameter. It is the way how `tflite` handles activation. We used similar (or same) kernel if there is no reason to do differently. Thus, handling differently only in `MaxPool`'s `forward` may make others confused. I prefer to keep the same style.

File: runtime/onert/backend/train/ops/PoolLayer.cc (hunk 7 above)
Comment (glistening) [reply runtime/onert/backend/train/ops/PoolLayer.cc:106]: I like the similar structure between `forward` and `backward` in same operator. But for me, keeping `forward` consistent to existing inference code as much as possible is more important. Please find more https://github.com/Samsung/ONE/pull/11686/files#r1353573206.

File: compute/cker/include/cker/train/operation/MaxPool.h (hunk 9 above)
Comment (zetwhite) [reply compute/cker/include/cker/train/operation/MaxPool.h:102]: @glistening Oh, I fully understood your intention. I'll move activation part to cker.

File: runtime/onert/backend/train/ops/PoolLayer.cc (hunk 7 above)
Comment (ragmani) [reply runtime/onert/backend/train/ops/PoolLayer.cc:106]: You can use `cwiseMin()` and `cwieMax()` of `out_mat` in MaxPool.h of cker like below.
https://github.com/Samsung/ONE/blob/31fc7a8267e38cd0879e85b8e6f4cfd942a010f7/compute/cker/include/cker/eigen/eigen_gemm_eigen.h#L84-L87

File: compute/cker/include/cker/train/operation/MaxPool.h (hunk 5 above)
Comment (zetwhite) [reply compute/cker/include/cker/train/operation/MaxPool.h:42]: updated!

File: compute/cker/include/cker/train/operation/MaxPool.h (hunk 9 above)
Comment (zetwhite) [reply compute/cker/include/cker/train/operation/MaxPool.h:105]: updated!

File: runtime/onert/backend/train/KernelGenerator.h (hunk 2 above)
Comment (zetwhite) [reply runtime/onert/backend/train/KernelGenerator.h]: Oh, I see. updated!

File: runtime/onert/backend/train/ops/PoolLayer.cc (hunk 6 above)
Comment (zetwhite) [reply runtime/onert/backend/train/ops/PoolLayer.cc]: updated

File: compute/cker/src/train/MaxPool.test.cc [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+  int b;
+  int h;
+  int w;
+  int d;
+};
+
+} // namespace
+
+TEST(CKer_Operation, MaxPool2D_Depth1)
```
Comment (zetwhite) [inline compute/cker/src/train/MaxPool.test.cc:42]: Test Results

```bash
➜ Product/out/unittest/test_cker --gtest_filter='*MaxPool*'  
Running main() from /home/sseung/Workspace/ONE/externals/GTEST/googletest/src/gtest_main.cc
Note: Google Test filter = *MaxPool*
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from CKer_Operation
[ RUN      ] CKer_Operation.MaxPool2D_Depth1
[       OK ] CKer_Operation.MaxPool2D_Depth1 (0 ms)
[ RUN      ] CKer_Operation.MaxPool2D_Depth2
[       OK ] CKer_Operation.MaxPool2D_Depth2 (0 ms)
[----------] 2 tests from CKer_Operation (0 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (0 ms total)
[  PASSED  ] 2 tests.
```

File: compute/cker/include/cker/train/operation/MaxPool.h [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+            const int out_offset = NodeOffset(b, ph, pw, output_height, output_width);
+            const int in_offset = NodeOffset(b, h, w, input_height, input_width);
+
+            const auto out_vector = out_mat.col(out_offset);
+            const auto in_vector = in_mat.col(in_offset);
+
+            // update arg_max_index_mat
+            arg_max_index_mat.col(out_offset) =
+              (out_vector.array() < in_vector.array())
+                .select(in_offset, arg_max_index_mat.col(out_offset));
```
Comment (zetwhite) [inline compute/cker/include/cker/train/operation/MaxPool.h:92]: (notes) 

eigen `select` syntax from (  https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#a65e78cfcbc9852e6923bebff4323ddca )

```
(condition).select(if-selected, else-selected);
```
```cpp
MatrixXi m(3, 3);
m << 1, 2, 3,
     4, 5, 6,
     7, 8, 9;
m = (m.array() >= 5).select(-m, m);
cout << m << endl;

// output : 
// 1 2 3
// 4 -5 -6
// -7 -8 -9
```

**Comment by ragmani** [comment]
Could you add negative tests? We should have more negative tests than positive tests.

File: compute/cker/include/cker/train/operation/MaxPool.h (hunk 8 above)
Comment (zetwhite) [reply compute/cker/include/cker/train/operation/MaxPool.h:55]: I added here - https://github.com/Samsung/ONE/issues/11248#issuecomment-1756798349

**Comment by zetwhite** [comment]
> Could you add negative tests? 

Oh I see. I'll add more negative (also padding) tests.

**Comment by zetwhite** [comment]
> Oh I see. I'll add more negative (also padding) tests.

(Progress updates) While working on this, i found some suspicious part with a padding example. 
That's why it took some time.  :cry:

File: compute/cker/src/train/MaxPool.test.cc [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+    MaxPoolOpVerifier<float> verifier(op_param, in, out);
+
+    std::vector<float> input = {0, 0, 0, 0};
+    std::vector<float> expected_output = {0, 0, 0, 0};
+    verifier.verifyForward(input, expected_output);
+
+    std::vector<float> output_deriv = {0.1, 0.1, 0.1, 0.2};
+    std::vector<float> expected_input_deriv = {0.1, 0.1, 0.1, 0.1};
+    verifier.verifyBackward(output_deriv, expected_input_deriv, false);
+  }
```
Comment (zetwhite) [inline compute/cker/src/train/MaxPool.test.cc:278]: (lately updated) Here, I updated tc with padding and negative one. (https://github.com/Samsung/ONE/pull/11686#issuecomment-1757001754)

**Review by ragmani — APPROVED** [review]
LGTM

**Review by glistening — APPROVED** [review]
LGTM

**Review by jyoungyun — APPROVED** [review]
LGTM
